### PR TITLE
BC PR B: cross-entropy loss + BcTrainer epoch loop + smoke test

### LINKED ISSUE #167: BC PR B: cross-entropy loss + BcTrainer epoch loop + smoke test
**Part of #161** (BC epic). **Depends on #164** (Demonstrations dataset + BcConfig). Labeled `loom:blocked` until #164 merges.

## Background
With the `Demonstrations` dataset and `BcConfig` in place (#164), this PR adds the supervised cross-entropy loss and the `BcTrainer` epoch loop — the core of Behavioral Cloning. Per epic decision 3 (#161), BC reuses `MlpBurnPolicy` (training only its action/logits head via cross-entropy; the value head is ignored), `BurnOptimizer`, and PPO's public `generate_minibatch_indices` for the seeded shuffle. It DIVERGES from every RL trainer in loop shape: a supervised epoch loop over a fixed dataset, no environment interaction, no advantages/returns/entropy. This is PR B of the BC epic.

### 1. `bc/loss.rs` — supervised cross-entropy
```rust
/// -mean(log_softmax(logits).gather(actions)) — NLL of the expert action.
pub fn compute_bc_loss<B: Backend>(logits: Tensor<B, 2>, actions: Tensor<B, 1, Int>) -> Tensor<B, 1>;
```
Discrete-only (matches `MlpBurnPolicy`). Re-use `scalar_f64` (re-exported via `a2c::loss` / `ppo::loss`) for host-side stat extraction.

### 2. `bc/trainer.rs` — `BcTrainer` + `BcEpochStats`
```rust
pub struct BcEpochStats { pub loss: f64, pub accuracy: f64 }

pub struct BcTrainer<B, P, O> { /* config, Option<P>, optimizer, rng, total_steps, total_epochs */ }

impl<B, P, O> BcTrainer<B, P, O> {
    pub fn new(config: BcConfig, policy: P, optimizer: BurnOptimizer<B, P, O>) -> Result<Self>;
    pub fn config(&self) -> &BcConfig;
    pub fn policy(&self) -> &P;
    pub fn total_steps(&self) -> usize;   // gradient updates (one per minibatch)
    pub fn total_epochs(&self) -> usize;
    /// Shuffle (seeded) indices, iterate minibatches of `batch_size`, compute
    /// cross-entropy, backprop, step the optimizer once per minibatch. Returns
    /// mean loss + mean action-match accuracy over the epoch.
    pub fn train_epoch<F>(&mut self, demos: &Demonstrations, forward_fn: F) -> Result<BcEpochStats>
    where F: FnMut(&P, Tensor<B, 2>) -> Tensor<B, 2>; // (policy, obs) -> logits
}
```
- Use the same `Option<P>` move-through-`step` ownership pattern as `A2cTrainer`/`PPOTrainerBurn` (Burn's `Optimizer::step` consumes the module by value).
- Shuffle via PPO's `generate_minibatch_indices` (or an equivalent seeded shuffle driven by the trainer's `StdRng`), so two runs with the same `seed` produce identical minibatch order.
- Accuracy = fraction of minibatch examples whose argmax-logit action matches the expert label, averaged over the epoch.
- `forward_fn` lets the caller pick how logits are produced (e.g. `|p, o| p.forward(o).0` to drop the value head).

### 3. Always-on fast smoke test (`#[cfg(test)]` in `bc/trainer.rs`)
On `Autodiff<NdArray<f32>>`: build a tiny synthetic `Demonstrations` (a handful of `(obs, action)` rows), construct a `BcTrainer`, run ONE `train_epoch`, assert (a) `total_steps() >= 1`, (b) `stats.loss.is_finite()`, (c) `stats.accuracy.is_finite()` and in `[0, 1]`. Fast, deterministic, no feature gate beyond `training`.

### 4. Module wiring
Add `pub mod loss;` and `pub mod trainer;` to `src/train/bc/mod.rs` with re-exports (`BcTrainer`, `BcEpochStats`, `compute_bc_loss`); extend the `src/train/mod.rs` re-exports.

## Acceptance criteria
- `compute_bc_loss` returns a finite scalar; a unit test confirms loss decreases when logits are pushed toward the expert labels (sanity, mirroring `a2c/loss.rs` loss-direction tests).
- `BcTrainer::train_epoch` performs `ceil(len / batch_size)` gradient steps per epoch and returns finite `loss` + `accuracy in [0, 1]`.
- Fast smoke test passes on every `cargo test --features training` run.
- Two `BcTrainer`s built with the same `BcConfig::seed` and same demos produce identical first-epoch stats (reproducibility).
- Zero edits to existing trainers (PPO/DQN/SAC/A2C).

## Affected files
- NEW `src/train/bc/loss.rs`, `src/train/bc/trainer.rs`
- EDIT `src/train/bc/mod.rs`, `src/train/mod.rs` (re-exports)

## References
- Trainer pattern (ownership, `Option<P>`, optimizer step, seeded RNG, stats struct): `src/train/a2c/trainer.rs`.
- Loss pattern + `scalar_f64` re-export: `src/train/a2c/loss.rs`, `src/train/ppo/loss.rs`.
- Logits/log_softmax/gather: `src/policy/mlp.rs` (`forward`, `evaluate_actions`).
- Seeded minibatch shuffle: `generate_minibatch_indices` (re-exported from `src/train/ppo`).

### PR DESCRIPTION
**Pull request by rjwalters**
## Summary

PR B of the Behavioral Cloning epic (#161), building on the merged `Demonstrations` dataset + `BcConfig` (#164). Adds the supervised cross-entropy loss and the `BcTrainer` epoch loop — the core of BC. Mirrors the A2C module layout (`src/train/a2c/{loss,trainer}.rs`).

### `src/train/bc/loss.rs` (new) — supervised cross-entropy
`compute_bc_loss(logits: [batch, action_dim], actions: [batch]) -> [1]`:
```text
loss = -mean_i( log_softmax(logits_i)[action_i] )
```
`log_softmax` over the action dimension, `gather` the log-prob at each expert action label (labels reshaped `[batch] -> [batch, 1]` for `gather`), negate the batch mean. This is the standard discrete classification cross-entropy / NLL specialized to `MlpBurnPolicy`'s categorical action head; the value head is ignored. Re-exports `scalar_f64` from `ppo::loss` (same pattern A2C uses).

### `src/train/bc/trainer.rs` (new) — `BcTrainer` + `BcEpochStats`
Supervised epoch loop over `Demonstrations`. DIVERGES from RL trainers: no env interaction, no advantages/returns/entropy, many gradient steps per call. Per `train_epoch`:
1. Seeded shuffle of example indices into minibatches of `config.batch_size` via PPO's `generate_minibatch_indices_with_rng` (driven by the trainer's `StdRng`, seeded from `config.seed`).
2. Per minibatch: gather `(obs, actions)`, produce logits via caller `forward_fn` (e.g. `|p, o| p.forward(o).0` to drop the value head), compute cross-entropy, backprop, step the optimizer once.
3. Returns example-weighted mean loss + argmax-match accuracy.

Reuses `BurnOptimizer` and the `Option<P>` move-through-`step` ownership pattern from `A2cTrainer`/`PPOTrainerBurn`.

### Module wiring (additive)
`pub mod loss; pub mod trainer;` + re-exports in `src/train/bc/mod.rs`; `BcEpochStats`, `BcTrainer`, `compute_bc_loss` re-exported from `src/train/mod.rs`. Zero edits to existing trainers.

## Tests (fast, always-on)
- Loss: finite scalar; non-negative; loss decreases as logits favor the expert label; ~0 when confidently correct.
- Trainer: constructs at zero steps; **smoke** — one epoch over a tiny 6-example set does `ceil(6/4)=2` steps and yields finite loss + accuracy in `[0,1]`; reproducibility — same seed + demos -> identical first-epoch stats; empty dataset rejected.

The slow CartPole cloned-vs-expert eval test is intentionally deferred to #169.

## Test Plan
- `cargo fmt --check`: clean
- `cargo test --features training --lib bc`: 15 passed (8 new), 0 failed
- `RUSTDOCFLAGS="-D warnings" cargo doc --no-deps --all-features`: clean
- `cargo clippy --all-targets --features training -- -D warnings`: zero errors in touched `src/train/bc/` files

## Pre-existing Issues (Not Addressed)

Repo-wide `cargo clippy -- -D warnings` reports ~56 pre-existing errors in unrelated files (`src/buffer/rollout/*`, `src/env/games/*`, `src/train/dqn/*`, `src/utils/normalize.rs`, etc.), present on `main` before this branch. None are in the files this PR touches; out of scope here.

Closes #167

### CODE REVIEW

**Comment by rjwalters** [comment]
## Judge Review: APPROVED

Verified PR B of the BC epic (#161). Additive-only: new `src/train/bc/{loss,trainer}.rs` plus module re-exports; zero edits to existing trainers.

### Loss math — correct
`compute_bc_loss(logits [batch, action_dim], actions [batch] Int)`:
- `log_softmax(logits, 1)` — log-probs over the action dim (dim 1).
- `actions.reshape([batch, 1])` then `log_probs.gather(1, idx)` — picks the log-prob at each expert label, `[batch, 1]`.
- `.reshape([batch]).mean().neg()` — mean NLL, negated.

This is exactly `-mean(log_softmax(logits).gather(action_labels))`, the standard discrete-classification cross-entropy specialized to MlpBurnPolicy's categorical head; value head ignored. Directional 'loss decreases when logits favor the expert label' test present (`bc_loss_decreases_when_logits_match_labels`), plus non-negativity and ~0-when-perfect.

### BcTrainer — correct
- Seeded `StdRng` from `config.seed`; per-epoch shuffle via PPO's `generate_minibatch_indices_with_rng`; `ceil(len/batch_size)` steps/epoch.
- One gradient step per minibatch; reuses `BurnOptimizer` + the `Option<P>` move-through-`step` ownership pattern (mirrors A2cTrainer).
- `forward_fn` (`|p, o| p.forward(o).0`) drops the value head.
- Example-weighted mean loss + argmax-match accuracy; empty dataset rejected with `Err`.

### Smoke test — correct
Fast always-on: tiny 6-example set, one epoch, asserts 2 steps + finite loss + accuracy in [0,1]. Reproducibility test (same seed -> identical stats). No slow CartPole eval test (correctly deferred to #169).

### Local verification (worktree at PR head 9fb3b43)
- `cargo test --features training --lib bc`: 15 passed, 0 failed (8 new)
- `cargo fmt --check`: clean
- `cargo clippy --lib --features training`: 0 warnings (touched bc files clean)
- `RUSTDOCFLAGS=-D warnings cargo doc --no-deps --features training`: clean

### CI
Fast checks all green (Rustfmt, Check (no training) x2, Documentation, Security Audit). Full Tests CI (Linux/macOS NdArray, ~1-2h) pending — not blocking per CI policy; local lib tests cover the new code.

Minor (non-blocking): per-minibatch host-side accuracy via `into_data().to_vec()` adds slight overhead — fine for correctness.

Approving.
